### PR TITLE
[plugin.video.southpark_unofficial] v0.5.0

### PR DESCRIPTION
**Pull request by wargio**
### Description
Refactoring whole code and better performance and stability on video loading
### Checklist:
- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

**Comment by wargio** [comment]
i've changed the forum link to https due the warning given by travis.

File: plugin.video.southpark_unofficial/southpark.py [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+		self.helper  = SP_Helper (self.options  )
+		self.paths   = SP_Paths  (self.addon_id )
+
+	def notify(self, text, time=WARNING_TIMEOUT_SHORT):
+		utext = _encode(text)
+		uaddonname = _encode(self.addon_obj.getAddonInfo('name'))
+		utime = _encode(str(time))
+		uicon = _encode(self.paths.PLUGIN_ICON);
+		notification = 'Notification(%s, %s, %s, %s)' % (uaddonname, utext, utime, uicon)
+		xbmc.executebuiltin(notification)
```
Comment (Rechi) [inline plugin.video.southpark_unofficial/southpark.py]:
```suggestion
		xbmcgui.Dialog().notification(uaddonname, utext, uicon, utime)
```

File: plugin.video.southpark_unofficial/southpark.py [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+			self.notify(self.i18n.WARNING_BANNED_EPISODE, WARNING_TIMEOUT_LONG)
+		elif kodi.PARAM_MODE == PLUGIN_MODE_PREMIERE:
+			self.notify(self.i18n.WARNING_PREMIERE, WARNING_TIMEOUT_LONG)
+		elif kodi.PARAM_MODE == '':
+			self.create_menu()
+		else:
+			self.notify("---ERROR---")
+			return
+		xbmcplugin.endOfDirectory(self.phandle)
+		xbmc.executebuiltin('Container.SetViewMode(504)')
```
Comment (Rechi) [inline plugin.video.southpark_unofficial/southpark.py]: Setting the view mode isn't allowed for plugins.

File: plugin.video.southpark_unofficial/southpark.py [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+		self.PARAM_MODE         = urllib.unquote_plus(params.get('mode'     , ''))
+		self.PARAM_URL          = urllib.unquote_plus(params.get('url'      , ''))
+		self.PARAM_EP_TITLE     = urllib.unquote_plus(params.get('title'    , ''))
+		self.PARAM_EP_THUMBNAIL = urllib.unquote_plus(params.get('thumbnail', ''))
+
+	def debug(self):
+		log_debug("PARAM_MODE:         {0}".format(self.PARAM_MODE        ))
+		log_debug("PARAM_URL:          {0}".format(self.PARAM_URL         ))
+		log_debug("PARAM_EP_TITLE:     {0}".format(self.PARAM_EP_TITLE    ))
+		log_debug("PARAM_EP_THUMBNAIL: {0}".format(self.PARAM_EP_THUMBNAIL))
```
Comment (Rechi) [inline plugin.video.southpark_unofficial/southpark.py]: reduce the log spamming, by at least combining those into one log line

File: plugin.video.southpark_unofficial/southpark.py [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+		self.addon = addon
+		self.GEO_LOCATIONS   = ["US","UK","ES","DE","IT"]
+		self.AUDIO_AVAILABLE = ["en","es","de"]
+		self.VIDEO_QUALITY   = ["high","low"]
+
+	def debug(self):
+		log_debug("OPTIONS Geolocation          {0}".format(self.geolocation(True)))
+		log_debug("OPTIONS Audio                {0}".format(self.audio(True)))
+		log_debug("OPTIONS Show Subtitles       {0}".format(self.show_subtitles()))
+		log_debug("OPTIONS Play Random Directly {0}".format(self.playrandom()))
```
Comment (Rechi) [inline plugin.video.southpark_unofficial/southpark.py]: reduce the log spamming

File: plugin.video.southpark_unofficial/southpark.py [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+			# sp.de returns a JS instead of a JSON so i need to convert it
+			jsonrsp = jsonrsp.decode('utf-8')
+		seasonjson = _json.loads(jsonrsp)
+		for episode in seasonjson['results']:
+			self.add_episode(episode)
+
+	def play_episode(self, url, title, thumbnail):
+		log_debug(url)
+		log_debug(title)
+		log_debug(thumbnail)
```
Comment (Rechi) [inline plugin.video.southpark_unofficial/southpark.py]: reduce the log spamming

File: plugin.video.southpark_unofficial/southpark.py (hunk 1 above)
Comment (wargio) [reply plugin.video.southpark_unofficial/southpark.py]: fixed in commit (thanks!!)

File: plugin.video.southpark_unofficial/southpark.py (hunk 2 above)
Comment (wargio) [reply plugin.video.southpark_unofficial/southpark.py]: fixed in new pushed commit

File: plugin.video.southpark_unofficial/southpark.py (hunk 3 above)
Comment (wargio) [reply plugin.video.southpark_unofficial/southpark.py]: i've commented the line 571 & 572, so they will never be called.

File: plugin.video.southpark_unofficial/southpark.py (hunk 5 above)
Comment (wargio) [reply plugin.video.southpark_unofficial/southpark.py]: i've commented the line 571 & 572, so they will never be called.

File: plugin.video.southpark_unofficial/southpark.py (hunk 4 above)
Comment (wargio) [reply plugin.video.southpark_unofficial/southpark.py]: i've commented the line 571 & 572, so they will never be called.

File: plugin.video.southpark_unofficial/southpark.py (hunk 2 above)
Comment (wargio) [reply plugin.video.southpark_unofficial/southpark.py]: removed

**Pull request merged**